server/src/main/java/org/apache/iotdb/db/metadata/mtree/traverser/Traverser.java

```diff
@@ -49,17 +50,18 @@
  *   <li>collector: to collect customized results of the matched node or measurement
  * </ol>
  */
-public abstract class Traverser<R> extends AbstractTreeVisitor<IMNode, R> {
+public abstract class Traverser<R, N extends IMNode<N>> extends AbstractTreeVisitor<N, R> {
 
   private static final Logger logger = LoggerFactory.getLogger(Traverser.class);
 
-  protected IMTreeStore store;
+  protected IMTreeStore<N> store;
 
-  protected IMNode startNode;
+  protected N startNode;
   protected String[] nodes;
 
   // measurement in template should be processed only if templateMap is not null
   protected Map<Integer, Template> templateMap;
+  protected IMNodeFactory<N> nodeFactory;
```

Comment:
This can be pushed down to MeasurementTraverser. This class will act as a common traverser and will be placed in package node-commons.